### PR TITLE
[consensus] enable tcb in a separate execution domain (thread)

### PR DESCRIPTION
**Pull request by davidiw**
- If we are using network sockets, we need the entire safety rules api to be mutable, so this is updated
- add a lightweight networking stack for TCB purposes
- add a threading model to simulate moving TCB into its own environment
- add tests to verify it all works
- update the config so we can try it out in libra swarm / cluster test

### CODE REVIEW

File: consensus/safety-rules/src/network.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            let result = self.read_buffer();
+            if !result.is_empty() {
+                return Ok(result);
+            }
+        }
+    }
+
+    /// Blocking write until able to successfully send an entire message
+    pub fn write(&mut self, data: &[u8]) -> Result<()> {
+        let data_len = data.len() as u32;
```
Comment (zekun000) [inline consensus/safety-rules/src/network.rs]: is this a safe cast?

File: consensus/safety-rules/src/remote.rs [hunk 2]
```diff
@@ -56,17 +56,21 @@ impl<T: Payload> RemoteService<T> {
...
 
 pub struct RemoteClient<T> {
-    service: Arc<RwLock<RemoteService<T>>>,
+    service: Box<dyn TRemoteClient<T>>,
 }
 
 impl<T: Payload> RemoteClient<T> {
-    pub fn new(service: Arc<RwLock<RemoteService<T>>>) -> Self {
+    pub fn new(remote_service: Arc<RwLock<RemoteService<T>>>) -> Self {
+        let service = Box::new(LocalService { remote_service });
```
Comment (zekun000) [inline consensus/safety-rules/src/remote.rs]: i'm confused by those naming, `LocalService { remote_service}`

Comment (davidiw) [reply consensus/safety-rules/src/remote.rs]: Ahh, we should rename this service into SerializationService then the container could either be a Local (in memory) or Thread / Process for remote.

File: consensus/safety-rules/src/network.rs (hunk 1 above)
Comment (davidiw) [reply consensus/safety-rules/src/network.rs]: rewrote it to first ensure that it is within bounds and then cast.

File: consensus/safety-rules/src/remote.rs (hunk 2 above)
Comment (davidiw) [reply consensus/safety-rules/src/remote.rs]: renamed remote to serializer but did it in a new commit (appended to the end of this PR) as the Remote is in an earlier stack.

File: consensus/safety-rules/src/network.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            "Found data that is too large to decode: {}",
+            data.len()
+        );
+        let data_len = data.len() as u32;
+        self.write_all(&data_len.to_le_bytes())?;
+        self.write_all(data)?;
+        Ok(())
+    }
+
+    fn read_buffer(&mut self) -> Vec<u8> {
```
Comment (dmitri-perelman) [inline consensus/safety-rules/src/network.rs]: Wow, I would expect there would be some libraries that would do this kind of low level logic for reading from a buffer. Is it because you don't want TCB dependencies? (what these dependencies would look like?)

File: consensus/safety-rules/src/remote.rs [hunk 4]
```diff
@@ -103,3 +107,21 @@ impl<T: Payload> TSafetyRules<T> for RemoteClient<T> {
...
+
+pub trait TRemoteClient<T>: Send + Sync {
+    fn request(&mut self, input: SafetyRulesInput<T>) -> Result<Vec<u8>, Error>;
+}
+
+struct LocalService<T> {
+    pub remote_service: Arc<RwLock<RemoteService<T>>>,
+}
+
+impl<T: Payload> TRemoteClient<T> for LocalService<T> {
```
Comment (dmitri-perelman) [inline consensus/safety-rules/src/remote.rs]: Hey, sorry, I'm still confused by the structure. I guess we could either add comments or make the names more self-descriptive. To be clear, it's not about the design (I do understand we need some sort of remote proxy design pattern), it's about the mental load for future readers :).

File: consensus/safety-rules/benches/safety_rules.rs [hunk 5]
```diff
@@ -62,9 +56,45 @@ fn lsr(n: u64) {
...
+
+fn thread(n: u64) {
+    let signer = ValidatorSigner::from_int(0);
+    let file_path = NamedTempFile::new().unwrap().into_temp_path().to_path_buf();
+    let storage = OnDiskStorage::default_storage(file_path.clone()).unwrap();
+    let safety_rules_manager = SafetyRulesManager::new_local(storage, signer.clone());
+    lsr(safety_rules_manager.client(), signer, n);
+}
+
+pub fn benchmark(c: &mut Criterion) {
```
Comment (dmitri-perelman) [inline consensus/safety-rules/benches/safety_rules.rs]: Beautiful :) 🥇

File: consensus/safety-rules/src/persistent_storage.rs [hunk 6]
```diff
@@ -97,30 +115,50 @@ fn test_in_memory_storage() {
...
+        Self::new(file_path, false)
     }
 
     pub fn default_storage(file_path: PathBuf) -> Result<Box<dyn PersistentStorage>> {
-        if file_path.exists() {
-            return Self::new_storage(file_path);
+        Self::new(file_path, true)
+    }
+
+    fn new(mut file_path: PathBuf, default: bool) -> Result<Box<dyn PersistentStorage>> {
```
Comment (dmitri-perelman) [inline consensus/safety-rules/src/persistent_storage.rs]: It seems that this commit is independent from the rest of the changes in this PR. Could we move it into a separate PR? (unless I'm mistaken you can have it on top of master)

File: consensus/safety-rules/src/remote.rs (hunk 4 above)
Comment (davidiw) [reply consensus/safety-rules/src/remote.rs]: I believe this is fixed in a later commit but perhaps I should reorder them so that a natural flow of review is more ... accessible.

File: consensus/safety-rules/src/network.rs (hunk 3 above)
Comment (davidiw) [reply consensus/safety-rules/src/network.rs]: This is somewhat exploratory in context of what standard rust provides in their "pure" Networking code. we also need to add in better error handling, I should probably pepper this file with todos.

In context of other tools, we could consider mio, tokio, libp2p but each contains substantial amount of unsafe code and are rather large code bases. At least for the TCB, if we can build something that is moderately simple and reliable, we ought to go in that direction. We'll need to work with the networking team to ensure that whatever is ultimately done here is a subset of libranet.
